exp/etcdrestore/controllers/etcdmachinesnapshot_controller.go

```diff
@@ -60,18 +68,193 @@ func (r *EtcdMachineSnapshotReconciler) SetupWithManager(_ context.Context, mgr
+	log.Info("Starting deletion of EtcdMachineSnapshot", "name", etcdMachineSnapshot.Name)
+
+	// Perform any necessary cleanup of associated resources here
+	// Example: Delete associated snapshot resources
+	snapshotList := &snapshotrestorev1.EtcdMachineSnapshotList{}
+	if err := r.Client.List(ctx, snapshotList, client.InNamespace(etcdMachineSnapshot.Namespace)); err != nil {
+		log.Error(err, "Failed to list associated EtcdMachineSnapshots")
+		return ctrl.Result{}, err
+	}
+
+	for _, snapshot := range snapshotList.Items {
+		if snapshot.Spec.MachineName == etcdMachineSnapshot.Spec.MachineName {
+			if err := r.Client.Delete(ctx, &snapshot); err != nil {
+				log.Error(err, "Failed to delete associated EtcdMachineSnapshot", "snapshotName", snapshot.Name)
+				return ctrl.Result{}, err
+			}
+			log.Info("Deleted associated EtcdMachineSnapshot", "snapshotName", snapshot.Name)
+		}
+	}
+
+	// Remove the finalizer to allow deletion of the EtcdMachineSnapshot
+	controllerutil.RemoveFinalizer(etcdMachineSnapshot, snapshotrestorev1.ETCDMachineSnapshotFinalizer)
+	if err := r.Client.Update(ctx, etcdMachineSnapshot); err != nil {
+		log.Error(err, "Failed to remove finalizer from EtcdMachineSnapshot")
+		return ctrl.Result{}, err
+	}
+
+	// Log the completion of the deletion process
+	log.Info("Completed deletion of EtcdMachineSnapshot", "name", etcdMachineSnapshot.Name)
+
 	return ctrl.Result{}, nil
 }
+
+// checkSnapshotStatus checks the status of the snapshot creation process.
+func checkSnapshotStatus(ctx context.Context, r *EtcdMachineSnapshotReconciler, etcdMachineSnapshot *snapshotrestorev1.EtcdMachineSnapshot) error {
+	log := log.FromContext(ctx)
+
+	etcdSnapshotFileList := &k3sv1.ETCDSnapshotFileList{}
+
+	if err := r.Client.List(ctx, etcdSnapshotFileList); err != nil {
+		log.Error(err, "Failed to list ETCDSnapshotFile resources")
+		return err
+	}
+
+	patchBase := client.MergeFromWithOptions(etcdMachineSnapshot.DeepCopy(), client.MergeFromWithOptimisticLock{})
+
+	for _, snapshotFile := range etcdSnapshotFileList.Items {
+		_, _, snapshotName, readyToUse, err := extractFieldsFromETCDSnapshotFile(snapshotFile)
+		if err != nil {
+			log.Error(err, "Failed to extract fields from ETCDSnapshotFile")
+			continue
+		}
+
+		if readyToUse && snapshotName == etcdMachineSnapshot.Name {
+			etcdMachineSnapshot.Status.Phase = snapshotrestorev1.ETCDSnapshotPhaseDone
+			if err := r.Client.Status().Patch(ctx, etcdMachineSnapshot, patchBase); err != nil {
+				return fmt.Errorf("failed to update EtcdMachineSnapshot status to Done: %w", err)
+			}
+			return nil
+		}
+	}
+
+	// If no matching ready snapshot is found, return an error to retry
+	return errors.New("snapshot not ready")
+}
+
+// extractFieldsFromETCDSnapshotFile extracts the location, nodeName, snapshotName, and readyToUse fields from an ETCDSnapshotFile resource.
+func extractFieldsFromETCDSnapshotFile(snapshotFile k3sv1.ETCDSnapshotFile) (string, string, string, bool, error) {
```

Comment:
can this function be dropped now since we can access fields directly? Alternatively, it can be changed to only  validating the EtcdSnapshotFile.

Reply:
Changed the function to validate the EtcdSnapshotFile as suggested to check the fields before directly accessing them. Thanks for suggestion